core/arch/arm/kernel/secure_partition.c

```diff
@@ -0,0 +1,776 @@
+					    SP_MEM_ATTR_ACCESS_RW);
+		return true;
+	case FFA_SVC_MEMORY_ATTRIBUTES_SET_64:
+		service_compose_direct_resp(regs,
+					    sp_svc_set_mem_attr(regs->x4,
+								regs->x5,
+								regs->x6));
+		return true;
+	default:
+		EMSG("Undefined service id 0x%"PRIx32, (uint32_t)regs->x3);
+		service_compose_direct_resp(regs, SP_RET_INVALID_PARAM);
+		return true;
+	}
+}
+
+static bool stmm_handle_storage_service(struct thread_svc_regs *regs)
+{
+	uint32_t flags = TEE_DATA_FLAG_ACCESS_READ |
+		TEE_DATA_FLAG_ACCESS_WRITE |
+		TEE_DATA_FLAG_SHARE_READ |
+		TEE_DATA_FLAG_SHARE_WRITE;
+	TEE_Result res = TEE_SUCCESS;
+
+	switch (regs->x3) {
+	case FFA_SVC_RPMB_READ:
+		res = sec_storage_obj_read(TEE_STORAGE_PRIVATE_RPMB,
+					   (void *)regs->x4, regs->x5,
+					   regs->x6, flags);
+		service_compose_direct_resp(regs, res);
+
+		return true;
+	case FFA_SVC_RPMB_WRITE:
+		res = sec_storage_obj_write(TEE_STORAGE_PRIVATE_RPMB,
+					    (void *)regs->x4, regs->x5,
+					    regs->x6, flags);
+		service_compose_direct_resp(regs, res);
+
+		return true;
+	default:
+		EMSG("Undefined service id 0x%"PRIx32, (uint32_t)regs->x3);
+		service_compose_direct_resp(regs, SP_RET_INVALID_PARAM);
+		return true;
+	}
+}
+
+static bool spm_eret_error(int32_t error_code, struct thread_svc_regs *regs)
+{
+	regs->x0 = FFA_ERROR;
+	regs->x1 = FFA_PARAM_MBZ;
+	regs->x2 = (uint64_t)error_code;
+	regs->x3 = FFA_PARAM_MBZ;
+	regs->x4 = FFA_PARAM_MBZ;
+	regs->x5 = FFA_PARAM_MBZ;
+	regs->x6 = FFA_PARAM_MBZ;
+	regs->x7 = FFA_PARAM_MBZ;
+	return true;
+}
+
+static bool spm_handle_direct_req(struct thread_svc_regs *regs)
+{
+	uint16_t dst_id = regs->x1 & 0xFFFF;
+	/* Look-up of destination endpoint */
+	if (dst_id == mem_mgr_id)
+		return stmm_handle_mem_mgr_service(regs);
+	else if (dst_id == ffa_storage_id)
+		return stmm_handle_storage_service(regs);
+
+	EMSG("Undefined endpoint id 0x%"PRIx16, dst_id);
+	return spm_eret_error(SP_RET_INVALID_PARAM, regs);
+}
+
+static bool spm_handle_svc(struct thread_svc_regs *regs)
+{
+	switch (regs->x0) {
+	case FFA_VERSION:
+		EMSG("Received FFA version");
+		regs->x0 = FFA_VERSION;
+		return true;
+	case FFA_MSG_SEND_DIRECT_RESP_64:
+		IMSG("Received FFA direct response");
+		return return_helper(false, 0, regs);
+	case FFA_MSG_SEND_DIRECT_REQ_64:
+		IMSG("Received FFA direct request");
+		return spm_handle_direct_req(regs);
+	default:
+		EMSG("Undefined syscall 0x%"PRIx32, (uint32_t)regs->x0);
+		regs->x0 = 0;
+		regs->x1 = 1; //panic
+		regs->x2 = 0xabcd; //panic code
+		return false;
+	}
+}
+
+const struct tee_ta_ops secure_partition_ops __rodata_unpaged = {
+	.enter_open_session = stmm_enter_open_session,
+	.enter_invoke_cmd = stmm_enter_invoke_cmd,
+	.enter_close_session = stmm_enter_close_session,
+	.dump_state = sec_part_dump_state,
+	.destroy = sec_part_ctx_destroy,
+	.get_instance_id = sec_part_get_instance_id,
+	.handle_svc = spm_handle_svc,
+};
+
+void sec_part_save_return_state(struct thread_ctx_regs *ctx_regs,
```

Comment:
Unused function?

Reply:
Yes, this was part of a bigger patchset implementing more FFA stuff. I'll remove this